zerver/tornado/event_queue.py

```diff
@@ -1286,19 +1287,21 @@ def process_deletion_event(event: Mapping[str, Any], users: Iterable[int]) -> No
             if not client.accepts_event(event):
                 continue
 
+            deletion_event = dict(event)
```

Comment:
Why do we need this copy in addition to the `.copy` steps below?

Reply:
We have a `continue` statement case before that `.copy`. -- `Mapping` is read only type, we can't alter `event` directly.

Reply:
One improvement we can make in a separate small PR for:
```py
            for message_id in deletion_event["message_ids"]:
                compatibility_event = deletion_event.copy()
                compatibility_event["message_id"] = message_id
                del compatibility_event["message_ids"]
                client.add_event(compatibility_event)
```
to 
```py
            message_ids = deletion_event["message_ids"]
            del deletion_event["message_ids"]
            for message_id in message_ids:
                deletion_event["message_id"] = message_id
                client.add_event(deletion_event)
```

Reply:
I'm pretty sure that change would be a serious bug. If one just edits the same object, inserting it into different queues, then one will end up with many pointers to the same object (all with the last message ID at the end of the loop).  

I'm sure some refactor works here, but the one you propose seems dangerous.

Reply:
Made a small change to copy only if `topic == "" and not client.empty_topic_name`

```py
def process_deletion_event(event: Mapping[str, Any], users: Iterable[int]) -> None:
    for user_profile_id in users:
        for client in get_client_descriptors_for_user(user_profile_id):
            if not client.accepts_event(event):
                continue

            deletion_event = event
            if deletion_event.get("topic") == "" and not client.empty_topic_name:
                deletion_event = dict(event)
                deletion_event["topic"] = Message.EMPTY_TOPIC_FALLBACK_NAME

            # For clients which support message deletion in bulk, we
            # send a list of msgs_ids together, otherwise we send a
            # delete event for each message.  All clients will be
            # required to support bulk_message_deletion in the future;
            # this logic is intended for backwards-compatibility only.
            if client.bulk_message_deletion:
                client.add_event(deletion_event)
                continue

            for message_id in deletion_event["message_ids"]:
                compatibility_event = dict(deletion_event)
                compatibility_event["message_id"] = message_id
                del compatibility_event["message_ids"]
                client.add_event(compatibility_event)
```